libs/stats/odc/stats/io.py

```diff
@@ -209,6 +209,92 @@ def _ds_to_cog(self, ds: xr.Dataset, paths: Dict[str, str]) -> List[Delayed]:
             out.append(self._write_blob(cog_bytes, url, ContentType="image/tiff"))
         return out
 
+
+    def _get_single_band_thumbnail(self, ds: xr.Dataset, task: Task, single_band: Dict[str, str], input_geobox: GridSpec, odc_file_path: str) -> Delayed:
+        """
+        The single_band Dict can be:
+            1. single_band: {"measurement": "count_clear", "lookup_table": {"0":[150,150,110]}}, or
+            2. single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['blue']} or 
+            3. single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['red', 'green', 'blue']}
+        """
+        band = single_band['measurement']
+        thumbnail_path = odc_file_path.split('.')[0] + f"_{band}_thumbnail.jpg"
+        zero_band = numpy.zeros((task.geobox.shape[0], task.geobox.shape[1]))
+
+        # if use lookup_table to tune pixel, it will return [numpy.Array, numpy.Array, numpy.Array]
+        # if use bit to tune pixel, it will return numpy.Array
+        tuning_pixels, stretch = FileWrite()._filter_singleband_data(data=ds[band].values.reshape([task.geobox.shape[0], 
```

Comment:
Hi, @Kirill888.
How are you. Sorry to interrupt you again.

I already moved the thumbnail generation code to standalone methods. However, based on my understanding, this new method still have 'un-lazy' section like this line

> tuning_pixels, stretch = FileWrite()._filter_singleband_data(data=ds[band].values.reshape([task.geobox.shape[0], task.geobox.shape[1]])

Can I just added dask.delay to wrap them up and that's enough?

Reply:
> Can I just added dask.delay to wrap them up and that's enough?

probably not, because of xarray wrapping

see here for example that deals with both lazy and non-lazy inputs:

https://github.com/opendatacube/datacube-core/blob/develop/datacube/utils/cog.py